### DISCUSSION TITLE
NATS as a transport/protocol

### DISCUSSION THREAD

**Discussion by sourcec0de** [connectrpc/connect-go#363] [Q&A] [answered]
## Introduction
Hello, aspiring contributor here. :wave: 
First off, thanks for the hard work!
I'm enjoying `buf` `connect-web` and `connect-go`.

I'm not sure if this is the correct place for this discussion.
Feel free to transition this ticket to a Github discussion instead of an issue.

## Proposal
I am interested in adding NATs as a transport.

Here are some examples of gRPC/RPC over NATs in the wild.
1. [nats-grpc][nats-grpc]
2. [nrpc][nrpc]

## What does this solve? 
1. run gRPC services in private networks (no ingress) 
4. bridge security constrained networks (span multiple private clouds/on-prem)

## What I need to be successful
If adding NATS support is of interest, I would love to make this contribution.
I'm looking for some documentation around adding a new protocol/transport.
I initially thought to follow the pattern set by `protocol_grpc.go` and `protocol_connect.go`.
Please let me know if I'm looking in the right direction.

## Diagram
NATS bread and butter is enabling this type of communication topology.
![nats-multi-cloud](https://user-images.githubusercontent.com/1580651/191072850-f7f59601-cbdc-4a89-89d8-dcbd8d4fb81d.png)

[nats-grpc]: https://github.com/cloudwebrtc/nats-grpc
[nrpc]: https://github.com/nats-rpc/nrpc

**Comment by akshayjshah** [comment, marked as answer]
Hi @sourcec0de! This works today - you don't need to do anything :)

`protocol_connect.go` and `protocol_grpc.go` define two different RPC protocols, but both depend on HTTP as the transport. To use NATS, you just need a library that wraps NATS with a `net/http`-compatible abstraction. Luckily, https://github.com/ripienaar/nats-roundtripper already exists. Waldemar Quevedo (@wallyqs) from the NATS team has a commit demonstrating how to get this working end-to-end with connect-go: https://github.com/wallyqs/connect-go/commit/2e744ec4bf7ce3184f2f12b88c96103246f11184

**Comment by sourcec0de** [comment]
@wallyqs you're awesome!!!

@akshayjshah thank you for pointing this out!!!
This is EXACTLY what I was looking for.

With the example being so simple it might not be worth adding this to the package.
connect doesn't necessarily need to have NATS as a dependency in its go.mod.

:bow: thanks again!
